lib/collection/src/collection_manager/holders/proxy_segment.rs

```diff
@@ -703,27 +703,22 @@ impl SegmentEntry for ProxySegment {
         let deleted_indexes_guard = self.deleted_indexes.read();
         let created_indexes_guard = self.created_indexes.read();
 
-        if deleted_points_guard.is_empty()
+        let wrapped_version = self.wrapped_segment.get().read().flush(sync)?;
+        let write_segment_version = self.write_segment.get().read().flush(sync)?;
+
+        let is_all_empty = deleted_points_guard.is_empty()
             && deleted_indexes_guard.is_empty()
-            && created_indexes_guard.is_empty()
-        {
-            // Proxy changes are empty, therefore it is safe to flush write segment
-            // This workaround only makes sense in a context of batch update of new points:
-            //  - initial upload
-            //  - incremental updates
-            let wrapped_version = self.wrapped_segment.get().read().flush(sync)?;
-            let write_segment_version = self.write_segment.get().read().flush(sync)?;
-            let flushed_version = max(wrapped_version, write_segment_version);
-            *self.last_flushed_version.write() = Some(flushed_version);
-            Ok(flushed_version)
+            && created_indexes_guard.is_empty();
+
+        let flushed_version = if is_all_empty {
+            cmp::max(write_segment_version, wrapped_version)
         } else {
-            let _write_segment_version = self.write_segment.get().read().flush(sync)?;
-            // If intermediate state is not empty - that is possible that some changes are not persisted
-            Ok(self
-                .last_flushed_version
-                .read()
-                .unwrap_or_else(|| self.wrapped_segment.get().read().version()))
-        }
+            cmp::min(write_segment_version, wrapped_version)
+        };
```

Comment:
I'm not 100% sure if this is required, but my logic was:

If we have *no* data in `deleted_points`/`deleted_indexes`/etc, then there are no pending changes to `wrapped_segment`/`optimized_segment`. We can safely `ack` `write_segment_version`, because all changes were applied to `write_segment` only, `wrapped_segment` weren't affected at all.

If we *do* have in-memory data, it means some of the applied operations applied to proxy segment should have modified `wrapped_segment` (we just couldn't apply them, because `wrapped_segment` is locked by optimizer), so if we crash, we might want to reapply these operations to the `wrapped_segment`, so we ack `wrapped_version` (which should always be older than `write_segment_version`).

An example of second case is a simple `point_delete` operation, that would just delete the point from `wrapped_segment`. If we always ack `write_segment_version`, then after restart we will still have the point in `wrapped_segment`, but we already "acked" `point_delete` operation, so it won't be replayed on the `wrapped_segment`.